src/mango/src/mango_cursor.erl

```diff
@@ -44,27 +44,286 @@
+        ExcludedIndexes
+    ),
+    Candidates4 = enhance_candidates(Candidates3, ExcludedIndexesTags),
+    UnfavoredIndexesTags = tag_elems(
+        #{usable => true, reason => [unfavored_type], ranking => UnfavoredIndexesRank},
+        UnfavoredIndexes
+    ),
+    Candidates5 = enhance_candidates(Candidates4, UnfavoredIndexesTags),
+    NotChosenIndexesTags = tag_elems(
+        #{usable => true},
+        IndexesOfType
+    ),
+    Candidates6 = enhance_candidates(Candidates5, NotChosenIndexesTags),
+    NotUsableDetails =
+        lists:flatmap(
+            fun({Index, {Usable, Details}}) ->
+                case Usable of
+                    true -> [];
+                    false -> [{Index, Details}]
+                end
+            end,
+            UsabilityMap
+        ),
+    Candidates7 = enhance_candidates(Candidates6, NotUsableDetails),
+    WinnerColumnWidth =
+        case Winner of
+            none ->
+                0;
+            W ->
+                case mango_idx:columns(W) of
+                    X when is_list(X) -> length(X);
+                    all_fields -> 256
+                end
+        end,
+    {_, NotChosenDetails} =
+        lists:foldl(
+            fun({Index, _Prefix, Distance}, {N, Acc}) ->
+                ColumnWidth = length(mango_idx:columns(Index)),
+                Reason =
+                    case {} of
+                        _ when Distance > 0 -> [less_overlap];
+                        _ when ColumnWidth > WinnerColumnWidth -> [too_many_fields];
+                        _ -> [alphabetically_comes_after]
+                    end,
+                Tags = #{
+                    reason => Reason,
+                    ranking => N
+                },
+                {N + 1, [{Index, Tags} | Acc]}
+            end,
+            {0, []},
+            SortedIndexRanges
+        ),
+    Candidates8 = enhance_candidates(Candidates7, NotChosenDetails),
+    Candidates = maps:remove(Winner, Candidates8),
+    % produce the final list
+    ToList =
+        fun(Index, Tags, Acc) ->
+            #idx{type = IndexType} = Index,
+            #{usable := Usable, reason := Reason, ranking := Ranking} = Tags,
+            Covering =
+                case IndexType of
+                    <<"json">> -> mango_idx_view:covers(Index, Fields);
+                    _ -> null
+                end,
+            Reasons = [{[{name, hd(Reason)}]}],
+            Analysis =
+                {[
+                    {usable, Usable},
+                    {reasons, Reasons},
+                    {ranking, Ranking},
+                    {covering, Covering}
+                ]},
+            Entry =
+                {[
+                    {index, mango_idx:to_json(Index)},
+                    {analysis, Analysis}
+                ]},
+            [Entry | Acc]
+        end,
+    lists:reverse(maps:fold(ToList, [], Candidates)).
```

Comment:
Wondering, what is the reason for using `lists:reverse/1`? maps will traverse the elements in some unspecified order, maybe `lists:sort` would work better here?

(It's true that in most cases it turns out maps will sort its keys, especially for small map sizes when it just uses a list of KVs, but we shouldn't rely on it).

Reply:
Ahh, nice catch!  The use of `reverse` is just an old habit from Haskell, where `fold` is usually a fold right and the result should be reversed to the get back the original order.  But later (after writing this piece of code) I learned that folds are taken lighter (formally) in Erlang and they are usually fold lefts.

Independently of that, your comment about the original order is valid -- for what it is worth, the actual ordering does not matter.  The value in `analysis.ranking`, optionally combined with the reasons, determines the order and it is up to the client to sort (or not) the elements of the array before processing them.  I am not sure if the API shall offer any guarantees about that.

Reply:
For the record, it was agreed (off-line) with @willholley that the order of elements in the array do not matter.